src/GraphQL/Execution/DocumentExecuter.cs

```diff
@@ -213,10 +225,23 @@ await listener.AfterExecutionAsync(context)
 
             result.AddErrors(context.Errors);
         }
-        catch (OperationCanceledException) when (options.CancellationToken.IsCancellationRequested)
+        catch (OperationCanceledException) when (originalCancellationToken.IsCancellationRequested)
         {
+            // Re-throw the original cancellation exception when the the original CancellationToken is canceled
+            // (e.g. the client disconnects)
             throw;
         }
+        catch (OperationCanceledException) when (options.CancellationToken.IsCancellationRequested)
+        {
+            // If the operation was canceled due to a timeout, return a result with no data and an error
+            // or throw a TimeoutException based on the configuration
+            if (options.TimeoutAction == TimeoutAction.ThrowTimeoutException)
+                throw new GraphQLTimeoutException();
```

Comment:
Should we pass the original exception as an inner exception?

Reply:
Done